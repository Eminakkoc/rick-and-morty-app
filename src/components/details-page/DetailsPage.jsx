import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Page from 'components/page/Page';
import CharacterDetail from 'components/character-detail/CharacterDetail';
import LoadingCharacterDetail from 'components/character-detail/LoadingCharacterDetail';
import ErrorCharacterCard from 'components/character-card/ErrorCharacterCard';

import history from 'common/history/History';
import paths from 'routes/paths';

import queries from 'common/queries/queries';

import { graphql } from 'react-apollo';
import gql from 'graphql-tag';

const redirectToMainPage = () => {
  history.push({
    pathname: paths.MAIN_PAGE,
  });
};

class DetailsPage extends Component {
  componentWillMount() {
    const { selectedCharacter: { name } } = this.props;

    if (!name) {
      history.replace({
        pathname: paths.MAIN_PAGE,
      });
    }
  }

  componentDidUpdate(prevProps) {
    const { data: { refetch } } = this.props;
    const { selectedCharacter } = this.props;

    if (prevProps.selectedCharacter.name !== selectedCharacter.name) {
      refetch({
        name: selectedCharacter.name,
      });
    }
  }

  render() {
    const { data: { characters, error, loading } } = this.props;
    const results = characters && characters.results;
    let episodeList;
    let info;

    if (results && results[0]) {
      [info] = results;
      episodeList = results[0].episode
        && info.episode.map(episodeObject => episodeObject.name);
    }

    return (
      <Page backTransition={redirectToMainPage}>
        <div className="details-page-container">
          {
            error
            && (
              <ErrorCharacterCard />
            )
          }
          {
            !error && loading && <LoadingCharacterDetail />
          }
          {
            results && !error && !loading
            && (
              <CharacterDetail
                name={info.name}
                origin={info.origin.name}
                image={info.image}
                episodeList={episodeList}
              />
            )
          }
        </div>
      </Page>
    );
  }
}

DetailsPage.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool,
    error: PropTypes.shape(),
    characters: PropTypes.shape({
      results: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string,
        image: PropTypes.string,
        origin: PropTypes.shape({
          name: PropTypes.string,
        }),
        episode: PropTypes.arrayOf(PropTypes.shape({
          name: PropTypes.string,
        })),
      })),
    }),
    refetch: PropTypes.func.isRequired,
  }),
  selectedCharacter: PropTypes.shape({
    name: PropTypes.string,
    image: PropTypes.string,
  }),
};

DetailsPage.defaultProps = {
  data: {
    loading: false,
    error: false,
    characters: {
      results: [{
        name: '',
        image: '',
        origin: {
          name: '',
        },
        episode: {
          name: '',
        },
      }],
    },
  },
  selectedCharacter: {
    name: null,
    image: '',
  },
};

const query = gql`
${queries.getCharacterInfo()}
`;

export default graphql(query, {
  options: props => ({
    variables: {
      name: props.selectedCharacter && props.selectedCharacter.name,
    },
  }),
})(DetailsPage);
