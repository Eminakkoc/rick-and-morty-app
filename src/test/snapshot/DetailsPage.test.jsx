// MainPage.test.js
import React from 'react';
import DetailsPage from '../../components/details-page/DetailsPage';
import renderer from 'react-test-renderer';

test('Details Page Component Structure', () => {
  const component = renderer.create(
    <DetailsPage
      data={{
        refetch: () => {},
      }}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
