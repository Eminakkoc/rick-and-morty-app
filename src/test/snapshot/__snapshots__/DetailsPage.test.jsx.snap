// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page Component Structure 1`] = `
<div
  className="page-container"
>
  <div
    className="page-header"
  >
    <div
      className="page-logo"
    />
  </div>
  <button
    className="back-button"
    onClick={[Function]}
    type="button"
  >
    &lt; Back
  </button>
  <div
    className="details-page-container"
  />
</div>
`;
